Documentation restructuring and cleanup

nashif opened the pull request:
Resolves #12671

nashif committed:
doc: restructure Bluetooth docs

Put all realted docs in one page.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: bluetooth: split API docs

Split API docs into multiple pages.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: consolidated filesystem API reference

Fixed doxygen groups for filesystems and created one stop reference for
filesystems in the docs.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: display: cleanup display API page

Make the reference consistent with other APIs and fixed heading.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: add bluetooth to api reference

Add bluetooth API documentation to reference.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: changed structure and layout

Move to the new structure with both:

 - API Reference
 - User and Developer Guides

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: move release notes to own folder

Remove release note noise into a new folder.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: rename device driver section

Shorten title and remove redundant "device" from the title.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: device driver and dts cleanup

Clean up those sections and move them the API reference section.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: netowrking: split API reference

Split API reference into smaller pages per group.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
net: lwm2m: add  doxygen group

Add missing doxygen group to get included in documentation.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: networking: move API reference to new section

Move text from under subsystems and put it under API reference where it
makes sense.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: net: enable more net options in doxygen

Enable additional networking features to get them included in doxygen
output.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: move networking/bluetooth to user/developer guides

Include both networking and bluetooth in the user guides.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: move kconfig docs into one document

New functions section belongs in the overall Kconfig document.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: update known issues for new structure

Changed path for matching warnings.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
preview: https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/12991/docs/index.html

dbkinder reviewed (commented):
While I like the doc reorg, I'd like us to have an index.rst in every folder (a web standard).  So instead of reference.rst in the top level folder, I'd like to see an index.rst in the /api folder that's referenced by the top-level index.rst  (as talked about in issue #6776).  You agree?

nashif commented:
> While I like the doc reorg, I'd like us to have an index.rst in every folder (a web standard). So instead of reference.rst in the top level folder, I'd like to see an index.rst in the /api folder that's referenced by the top-level index.rst (as talked about in issue #6776). You agree?

yeah, thats why it is WIP. Right now I am just showing the structure I want, next step is moving files around and create index pages etc.

jhedberg commented on doc/subsystems/bluetooth/index.rst:
```diff
@@ -72,8 +76,8 @@ Bluetooth stack:
-PICS Features
-=============
+-PICS Features
+-=============
```
What's going on here? I.e. what are those `-` added to the beginning of the lines?

nashif replied:
sharp eyes, no idea how this happened.

jhedberg reviewed (dismissed):
Bluetooth changes look fine to me, except for the one issue I've commented on separately.

nashif committed:
doc: introduce final structure

Move guides and APIs into separate directories and cleanup naming
introducing index files rather than named section files.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@dbkinder new structure in place

nashif committed:
doc: change board dts documentation with new URLs

flash-partitions page has moved...

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented:
@nashif do you have this built somewhere so I can take a  look at the rendered result without having to build myself locally?

nashif committed:
doc: restructure and create index pages

Move all lead pages to be index pages and create redirect rules from the
old pointers.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
> @nashif do you have this built somewhere so I can take a look at the rendered result without having to build myself locally?

pointer in 2 comment

carlescufi reviewed (approved):
Just reviewed this side-by-side and I like the new layout

![image](https://user-images.githubusercontent.com/12450381/52232056-cfb12800-28bb-11e9-9cf7-ae2ba5b5bab1.png)

dbkinder reviewed (approved):
I like!  I've got a fix for getting the icons on the home page using fontawesome characters, so no licensing issues, once this gets merged.  (We can run a link checker afterward to see if links got missed.)

jukkar commented:
I wonder if it would be possible to change the API formatting a bit. IMHO it is easier to read the API doc if only the function name is highlighted with the color like in this example below. Having both the `Return` and `Parameter` highlighted makes it harder to read the doc. The Return and Parameter could be set in bold instead of color highlighting.

![example](https://user-images.githubusercontent.com/1098415/52259309-f490b400-292a-11e9-93c6-743fd56005ff.png)

pfalcon commented:
> Having both the `Return` and `Parameter` highlighted makes it harder to read the doc. 

Btw, as we discussed a couple of days ago on Slack with @nashif, it's weird that we have "Return" first, followed by "Parameters". We both admitted that we only now noticed that, so I don't know if it was always like that or some recent settings/them change. If somebody knows how to fix it, would be nice to do. 

> The Return and Parameter could be set in bold instead of color highlighting.

+1

nashif commented:
> I wonder if it would be possible to change the API formatting a bit.

open an issue please

nashif commented:
> t's weird that we have "Return" first, followed by "Parameters".

open an issue please, we can look at this later.
